Update everything to ES6 and adopt a code style

facekapow opened the pull request:
This PR is a big one, it changes pretty much all of the JavaScript files, since it updates everything to ES6, plus adopts [airbnb/javascript](https://github.com/airbnb/javascript) as a code style (with a few tweaks, like using require instead of ES6 modules).

Done so far:

- [x] `js/*.js`
- [x] `js/component`
  - [x] `js/component/dns-client`
- [x] `js/core`
  - [x] `js/core/*.js`
  - [x] `js/core/debug`
  - [x] `js/core/keyboard`
  - [x] `js/core/net`
  - [x] `js/core/pci`
  - [x] `js/core/ps2`
  - [x] `js/core/random`
  - [x] `js/core/stdio`
  - [x] `js/core/tty`
- [x] `js/driver`
  - [x] `js/driver/ps2`
  - [x] `js/driver/virtio`
      - [x] `js/driver/virtio/vring`
      - [x] `js/driver/virtio/*.js`
- [x] `js/modules`
- [x] `js/service`
  - [x] `js/service/dhcp-client`
  - [x] `js/service/dns-resolver`
  - [x] `js/service/shell`
- [x] `js/test`
  - [x] `js/test/unit`
    - [x] `js/test/unit/buffers`
    - [x] `js/test/unit/lib`
    - [x] `js/test/unit/net`
    - [x] `js/test/unit/platform`
    - [x] `js/test/unit/random`
    - [x] `js/test/unit/script`
    - [x] `js/test/unit/timers`
    - [x] `js/test/unit/virtio`
- [x] `js/utils`

vitkarpov commented:
Good buy commit history :)

facekapow commented:
Also added guidelines for contributing, plus a list of exceptions to Airbnb's style guide.

SpaceboyRoss01 commented:
How do we update runtime with npm.

facekapow commented:
This is done and ready for review/merge, the only thing I'm not really confident about is the `CONTRIBUTING.md`. I've got some more node compatibility fixes, but I'd prefer to add them in another PR after this has been merged to avoid any conflicts.

iefserge commented:
This is awesome, I'll do some tests too, to make sure everything works.

iefserge commented on js/component/dns-client/dns-packet.js:
```diff
@@ -70,98 +63,99 @@ exports.getQuery = function(domain, query) {
 
-function readHostname(reader) {
-  var labels = [];
+const readHostname = (reader) => {
```
I think regular function declarations are fine, stack trace will be nicer

iefserge commented on js/component/dns-client/dns-packet.js:
```diff
@@ -70,98 +63,99 @@ exports.getQuery = function(domain, query) {
-        label += String.fromCharCode(reader.readUint8());
-      }
+      let label = '';
+      for (let i = 0; i < len; ++i) label += String.fromCharCode(reader.readUint8());
```
I'd wrap single statement blocks into curly braces anyway, in my opinion it's easier to read. Single return statement is fine without braces though.

iefserge commented on js/core/net/ip4-address.js:
```diff
@@ -14,74 +14,63 @@
+  }
+  static get LOOPBACK() {
+    return new IP4Address(127, 0, 0, 1);
+  }
```
would be cool to have those ip objects allocated once and cached, getters can be slower too I guess

iefserge commented on js/core/net/mac-address.js:
```diff
@@ -14,59 +14,47 @@
+  }
+  static get ZERO() {
+    return new MACAddress(0, 0, 0, 0, 0, 0);
+  }
```
can be cached objects too

iefserge commented on js/core/net/tcp-socket-state.js:
```diff
@@ -13,14 +13,19 @@
+  'STATE_TIME_WAIT',
+];
+
+for (const num of Object.keys(states)) exports[states[num]] = num;
```
this would make it harder to convert code to ES6 exports, also numbers can be useful for debugging

iefserge commented on js/core/net/tcp-timer.js:
```diff
@@ -13,27 +13,19 @@
-  for (var connSocket of connections) {
-    connSocket._timerTick();
-  }
+let initTimeout;
```
function declaration is probably fine to use here

facekapow replied:
I just tried it and eslint errors that 'timeoutHandler' was used before it was defined... I could also just disable `no-use-before-define` if you prefer it strictly as a function declaration

iefserge replied:
ahh, interesting, I guess cases like that are pretty rare. I think either of those is fine:
- disable it for function declarations. http://eslint.org/docs/rules/no-use-before-define
- rewrite it using `setTimeout()` twice 

(not using `setInterval` here to prevent event queue overflow in case system can't process events fast enough)

iefserge commented on js/core/tty/vga.js:
```diff
@@ -43,89 +43,74 @@ var color = {
-  setCharOffset(this.b, offset, String(char), fg, bg);
-};
+// declare it up here, and define it below VGABuffer, a workaround for no-use-before-define
+let testInstance;
```
would function declaration work instead?

iefserge replied:
i.e https://github.com/airbnb/javascript#functions--declarations

facekapow replied:
Not really, same as above

iefserge commented on js/driver/virtio/net.js:
```diff
@@ -96,68 +92,55 @@ function initializeNetworkDevice(pciDevice) {
   if (!dev.writeGuestFeatures(features, driverFeatures, deviceFeatures)) {
-    debug('[virtio] driver is unable to start');
-    return;
+    return debug('[virtio] driver is unable to start');
```
nitpick, it shouldn't return debug() return value

iefserge commented on js/driver/virtio/net.js:
```diff
@@ -96,68 +92,55 @@ function initializeNetworkDevice(pciDevice) {
   transmitQueue.suppressUsedBuffers();
 
-  function fillReceiveQueue() {
+  const fillReceiveQueue = () => {
```
I'd prefer using function declarations instead of anonymous functions (where possible)

iefserge commented on js/driver/virtio/vring/descriptor-table.js:
```diff
@@ -32,30 +30,25 @@ class DescriptorTable {
-    }
+    let i;
+    for (i = 0; i < ringSize; ++i) this.descriptorsBuffers[i] = null;
+    for (i = 0; i < ringSize - 1; ++i) this.setNext(i, i + 1);
```
can use `let i` in each loop

iefserge commented on package.json:
```diff
@@ -21,9 +21,9 @@
   "devDependencies": {
-    "eslint": "^1.8.0",
-    "eslint-config-runtime": "^1.0.0",
-    "eslint-plugin-runtime-internal": "^1.0.0",
```
`eslint-plugin-runtime-internal` checks copyright headers in files, keep it?

iefserge commented on js/core/pci/scan.js:
```diff
@@ -780,89 +659,83 @@ pciManager.each(function(pciDevice) {
+
+  const info = `${address.bus.toString(16)}: ${address.slot.toString(16)}.${address.func}
+${pciDevice.vendorId().toString(16)}: ${pciDevice.deviceId().toString(16)}
+${classData.className} IRQ: ${vector} PIN: ${pins[devicePin]}`;
```
use spaces vs `\n` as separators

iefserge commented on js/__loader.js:
```diff
@@ -14,253 +14,209 @@
+            evalScriptFn(
+              `(function(require,exports,module,__filename,__dirname){
+${content}
+})((function(m){return function(path){return m.require(path)}})(global.module),global.module.exports,global.module,global.module.filename,global.module.dirname)`,
```
this shouldn't contain newlines to keep line numbers correct in stack traces

facekapow commented:
Ok, I've update the code to only use arrow functions for property functions and callbacks and function declarations for everything else, I've run the code through a beautifier, removed the newlines in `js/__loader.js`, used spaces instead of newlines in `js/core/pci/scan.js`, readded `eslint-plugin-runtime-internal`, disabled `no-use-before-define` for functions, and fixed the other issues you mentioned.

iefserge commented on js/driver/virtio/net.js:
```diff
@@ -17,7 +17,7 @@ const VirtioDevice = require('./device');
 const { MACAddress, Interface } = runtime.net;
 
-const virtioHeader = (() => {
+const virtioHeader = (function () { // eslint-disable-line wrap-iife
```
nitpick, arrow function if ok in this case?

iefserge commented on js/test/unit/net/ip4.js:
```diff
@@ -108,16 +131,42 @@ test('receive ip4 fragmented overlapped fragments', (t) => {
+  const slices = [{
+    offset: 8,
+    len: 64,
+  }, // [ ========]
```
looks ok, comment no longer readable though. Move it outside before the const?
or keep it as is if lint allows?

iefserge commented on js/test/unit/net/tcp-receive.js:
```diff
@@ -102,211 +110,389 @@ const sequenceNumbersList = [
+  receiveTestBatch(t, sequenceNumbersList, [{
+    seqOffset: 0,
+    len: 8,
+  }, {
```
minor q, airbnb doesn't allow props on the same line?

iefserge commented on js/test/unit/net/tcp.js:
```diff
@@ -273,11 +288,9 @@ test('tcp send FIN', (t) => {
-    socket._transmit = recvACKFIN;
-    socket.close();
-    t.equal(socket._state, tcpSocketState.STATE_TIME_WAIT);
+    socket._transmit = recvACKFIN; socket.close(); t.equal(socket._state, tcpSocketState.STATE_TIME_WAIT);
```
why on the same line?

iefserge commented on js/test/unit/net/tcp.js:
```diff
@@ -314,24 +327,16 @@ test('tcp receive FIN, then send more data', (t) => {
-    socket.close();
-    t.equal(socket._state, tcpSocketState.STATE_CLOSED);
+    socket._transmit = () => {}; socket.send(new Uint8Array([1, 2, 3]));
+    const packet = createTcpPacket(txSeq, rxSeq, tcpHeader.FLAG_FIN | tcpHeader.FLAG_ACK); socket._receive(packet, IP4Address.ANY, 45001, 0); t.equal(socket._state, tcpSocketState.STATE_CLOSE_WAIT); socket.send(new Uint8Array([4, 5, 6])); socket.send(new Uint8Array([7, 8, 9])); socket._transmit = handleLastAck; socket.close(); t.equal(socket._state, tcpSocketState.STATE_CLOSED);
```
single line?

facekapow commented:
Convert IIFE functions to arrow functions, and fixed the minor style issues. Sorry, most of these little things are because of the beautifier I ran it through.

iefserge commented on js/test/unit/net/tcp.js:
```diff
@@ -286,36 +290,30 @@ test('tcp send FIN', function(t) {
-    socket._transmit = recvACKFIN;
-    socket.close();
-    t.equal(socket._state, tcpSocketState.STATE_TIME_WAIT);
+    socket._transmit = recvACKFIN; socket.close(); t.equal(socket._state, tcpSocketState.STATE_TIME_WAIT);
```
single line?

iefserge commented:
@facekapow left couple of places where there are a few statements on the single line.

facekapow commented:
@iefserge Well, that's the last time I use a beautifier 😄

iefserge commented:
LGTM
great work 👍

kesla commented:
wow. Amazing.

facekapow commented:
Great, then I'll just do a quick once-over on the code and merge.

facekapow commented:
Everything looks ok, I tried to run the JS tests, but those are *definitely* broken (more specifically, `tape` kept complaining about `Stream`), but I did some manual feature testing and everything looks fine. Merge?

vitkarpov commented:
bye-bye diff :)

iefserge commented:
@facekapow yeah, I've been looking into tests issue, should be fine, going to fix it afterwards.
@vitkarpov yeah, but consistent and nice code style is good :)
